monotonic/src/main/scala/com/criteo/cuttle/monotonic/MonotonicScheduler.scala

```
@@ -0,0 +1,165 @@
+  extends Job[MonotonicScheduling[T]](id = triggerName, name = "trigger_job", description = "Triggers the execution of a workflow", tags = Set.empty, scheduling = mon)(_ => Future.successful(())) {
+  def apply() = trigger()
+}
+
+case class MonotonicScheduling[T: ClassTag: Encoder](trigger: MonotonicScheduling.Aux[T]) extends Scheduling {
+  type Context = MonotonicContext[T]
+  type DependencyDescriptor = MonotonicDependency
+
+  val self = this
+
+  val triggerJob = new TriggerJob("event", trigger, self)
+
+  implicit val scheduler: Scheduler[MonotonicScheduling[T]] = new MonotonicScheduler[T](triggerJob)
+}
+
+object MonotonicScheduling {
+  type Aux[T] = () => Future[T]
+}
+
+// TODO: maybe use a partial representation of the context to materialize timeouts, partial executions or filters
+// Example: half of the IDs were processed only from job A to job B
+trait MonotonicDependency
+
+case object MonotonicDependency extends MonotonicDependency
+
+class MonotonicScheduler[T: ClassTag: Encoder](trigger: TriggerJob[T], maxWorkflowParallelRuns: Int = 1) extends Scheduler[MonotonicScheduling[T]] {
+
+  type MonotonicJob = Job[MonotonicScheduling[T]]
+  // State is defined as a list of successful executions for each job
+  // We do not keep track of failed job in the state, they are listed as not executed
+  type State = Map[MonotonicContext[T], Set[MonotonicJob]]
+  val state = TMap.empty[MonotonicContext[T], Set[MonotonicJob]]
+
+  def addCompletedRuns(finishedExecutions: Set[(MonotonicJob, MonotonicContext[T])])
+                      (implicit txn: InTxn) =
+    finishedExecutions.foreach { case (job, context) =>
+      state.update(context,
+        state.get(context).getOrElse(Set.empty) + job)
+    }
+
+
+
+  def start(graph: Workflow[MonotonicScheduling[T]], executor: Executor[MonotonicScheduling[T]], xa: XA) = {
+    Database.doSchemaUpdates.transact(xa).unsafePerformIO
+
+    def runTrigger(): Unit =
+      trigger() onSuccess { case res =>
```

In the current implementation, a failure of the trigger function basically kill the scheduler forever as the failure is not handled.